### ISSUE TITLE
Unknown custom metadata item kind: 6 when using Visual Studio 2015 RTM

### ISSUE THREAD

**Issue by ca-ta** [intellifactory/websharper#402]
Websharper fails to compile F# projects that reference C# assemblies built with Roslyn, with the "unknown custom metadata item kind: 6" error.

I've tracked this down to Websharper  using Mono.Cecil 9.5.4, which has this issue. See [this](https://github.com/jbevain/cecil/issues/186) and [this](https://github.com/jbevain/cecil/issues/211).

The trivial solution would seem upgrading to Mono.Cecil **9.6.1**. 
However, I've given it a dry run and, with Mono.Cecil 9.6.1, Websharper compilation is significantly slower.

I'm wondering if you guys have any insights as to why this would happen (I mean the compile time degradation) and what are your plans regarding Visual Studio 2015.

Thanks!

**Comment by ca-ta** [comment]
Any plans regarding this? - it is effectively preventing me from upgrading to VS 2015 and F# 4.0

**Comment by granicz** [comment]
I don't understand how this is related to VS 2015, is it not about Roslyn-generated C# assemblies?  I assume you are using these to supply backend functionality that is called from RPCs, is that correct?

**Comment by ca-ta** [comment]
Thank you for replying. Yes, indeed, it is about Roslyn-generated C# assemblies. 
In my particular case, one of the projects in the solution *must* be a C#  project containing [Orleans](https://github.com/dotnet/orleans) interfaces - this is due to Orleans-specific code-gen requirements. Switching to VS 2015 causes said project to be build using Roslyn. The C# project is referenced from an F# server-side component that must also be compiled with WebSharper.  WebSharper compilation fails, due to the fact that Cecil 9.5.4 cannot deal with the Roslyn- generated assembly.

**Comment by ca-ta** [comment]
Hey guys, It would be great if you could share your plans regarding support projects referencing Roslyn-generated assemblies.
I am thinking Nuget packages including such assemblies will soon appear in the wild (as people move to VS 2015) - WebSharper would fail to build any assembly referencing such Nuget packages.

**Comment by ca-ta** [comment]
Hope this helps someone hitting this issue:
The WebSharper compilation failure is caused by Cecil not 9.5.4 being able to parse the PDB information generated by Roslyn. 
A stop-gap workaround is to change build settings for the C# projects that are referenced from assemblies that need to be build with WebSharper to *not* generate debugging information (Build/Advanced/Debug info: none).

**Comment by jhamm** [comment]
That is a good tip. Thanks!

**Comment by granicz** [comment]
Thanks for investigating this issue, the insights you gave have been tremendously helpful.  Were these reported on the Cecil or Roslyn forums?  At this point, I am not sure what to do inside WebSharper, other than waiting out Roslyn or Cecil to work properly?

**Comment by ca-ta** [comment]
@granicz  - I think it would be sufficient to *update the Cecil Nuget to 9.6.1*. 

See my original comment for links to the relevant Cecil issues (the issues have been addressed in 9.6.1).

I guess we'll have to see if it negatively impacts compilation times - as it was my experience back in May - and deal with that as a separate issue if it occurs.

I would have done that myself and issued a PR, but I am having trouble building the latest WebSharper on my machine and haven't found the time to figure out what's going on (I am using the build script with a clean get, but it just doesn't build). Anecdotally, building WebSharper has always been a finicky experience for me - It seems that the problems I've been experiencing have to do with the bootstrapper and the custom Nuget package management performed as part of the build (and yes, I've used the build script).

**Comment by ca-ta** [comment]
Hi guys, I have just stumbled upon the a Nuget package in the wild that brakes WebSharper:  `Serilog 1.5.11`
Referencing Serilog 1.5.11 from an F# assembly built with WebSharper yields the dreaded "Unknown custom metadata item kind: 7".

**Comment by Tarmil** [comment]
As I looked more into this, I found out that pdb-loading errors are already managed [here](https://github.com/intellifactory/websharper/blob/master/src/compiler/WebSharper.Compiler/api/Loader.fs#L99), just with too narrow a pattern so this exception gets through. I just tested replacing it with `_`, and I can compile a project that references `Serilog` with Cecil 9.5. This should be the best solution, at least until we find why Cecil 9.6 is so slow.
